### PR TITLE
POC/Discuss: Building integrations package

### PR DESCRIPTION
**Pull request by ruflin**
This is a proposal on how to build integrations packages for the integration registry (https://github.com/elastic/integrations-registry). The details of this proposal can be found in the `integration/README.md` file. The ideas is to comment on the proposal directly there, so changes can be  discussed per topic and changes are directly visible.

I expect this to be a long running and a PR with lots of discussion. The current content is a first stab at the proposal and we should keep iterating on it. The main goal of this is to make development of integrations as easy as possible and at the same time allow all the automation and verifications we have around integrations in place. Using the original package format and adding meta information would not allow to reuse some content (ECS fields as an example) and would make testing of each dataset separately harder. Each dataset can require a different test setup.

The apache and coredns integrations were taken as an example to have at least to example on how things could work. I added these as I guess discussing directly on the code and the generated content makes discussion easier.

The current code is not final, it has many TODO and duplicated code inside. So please do not review the code yet, but the concept.

The initial discussion around this started in this [google doc](https://docs.google.com/document/d/1OwTNEc0DTcSyPktwbi7RQPC7iEOh_pZPRyx500kixrg/edit#) but all discussions from there an information should make into here to have it in public.

To build the existing packages, run `mage build`.

### CODE REVIEW

File: integration/apache/dataset/apache.access/kibana/index-pattern/metricbeat.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
..._values\":true,\"indexed\":true,\"name\":\"docker.cpu.kernel.pct\",\"scripted\":false,\"searchable\":true,\"type\":\"number\"},{\"aggregatable\":true,\"analyzed\":false,\"count\":0,\"doc_values\":true,\"indexed\":true,\"name\":\"docker.cpu.kernel.ticks\",\"scripted\":false,\"searchable\":true,\"type\":\"number\"},{\"aggregatable\":true,\"analyzed\":false,\"count\":0,\"doc_values\":true,\"indexed\":true,\"name\":\"docker.cpu.system.pct\",\"scripted\":false,\"searchable\":true,\"type\":\"number\"},{\"aggregatable\":true,\"analyzed\":false,\"count\":0,\"doc_values\":true,\"indexed\":true,\"name\":\"docker.cpu.system.ticks\",\"scripted\":false,\"searchable\":true,\"type\":\"number\"},{\"aggregatable\":true,\"analyzed\":false,\"count\":0,\"doc_values\":true,\"indexed\":true,\"name\":\"docker.cpu.user.pct\",\"scripted\":false,\"searchable\":true,\"type\":\"number\"},{\"aggregatable\":true,\"analyzed\":false,\"count\":0,\"doc_values\":true,\"indexed\":true,\"name\":\"docker.cpu.user.ticks\",\"scripted\":false,\"searchable\":true,\"type\":\"number\"},{\"aggregatable\":true,\"analyzed\":false,\"count\":0,\"doc_values\":true,\"indexed\":true,\"name\":\"docker.cpu.total.pct\",\"scripted\":fal
```
Comment (jsoriano) [inline integration/apache/dataset/apache.access/kibana/index-pattern/metricbeat.json]: Metricbeat index pattern under this dataset?

File: integration/apache/dataset/apache.access/testing.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# For testing definitions
+
+service:
+  filebeat:
+    min: 7.0.0
+  elasticsearch: 7.0.0
+  kibana: 7.0.0
+  # No running version of apache needed for logs
```
Comment (jsoriano) [inline integration/apache/dataset/apache.access/testing.yml]: It may be interesting to support this for services that log to syslog

File: integration/apache/dataset/apache.access/testing.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# For testing definitions
```
Comment (jsoriano) [inline integration/apache/dataset/apache.access/testing.yml]: It'd be good to take this into acocunt https://github.com/elastic/beats/issues/10182

File: magefile.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build mage
```
Comment (jsoriano) [inline magefile.go]: I wonder if we should have an external tool instead of a `magefile.go` to ease development of new integrations out of this repository. This external tool could even be a beats subcommand.

File: integration/README.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+The manifest contains all the information about the integration and follows the logic of the [package manifest](https://github.com/elastic/integrations-registry/blob/master/ASSETS.md#general-assets). The manifest might be enriched with further information from its dataset during packaging. Also verifications on compatiblity version etc. will be done.
+
+It contains a few additional fields which are not part of the package:
+
+**datasets**
+
+This is a list of dataset this integration depends on. As packages today do not allow to depend on other packages, it is important to have a dependency feature during building integration to not have to duplicate all the assets. Some examples here are fields for ECS or fields specific to Filebeat. An example is below:
+
+```
+datasets:
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: I like `includes` as in the end it's included in the package.

File: integration/README.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    - name: "ecs:ecs"
+    - name: "filebeat:filebeat"
+```
+
+No versions are mentioned above of the datasets. It's up to the implementer to make sure to increase the version number of the integration if a dependency changes. Alternative we could use versions which are then validate and if not correct anymore, an error is thrown. This would probably be more dev friendly but more complex to implement.
+
+**Package config**
+
+Not all integrations which are in this repo need packaging. For example the Filebeat or ECS integration directory are only placeholders for the assets but will not come any integration. To prevent these from packaging. `package: false` can be set.
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: The assets of filebeat and ecs are copied over to each package that `includes` it.

I was thinking about packages depending on packages but decided to not support it for now. With this, we reduce the complexity and dependency hell on install time. We can't have circular dependencies, we know all versions are around etc. It needs a bit more tooling when creating the packages but we always now, the package we get is self sufficient and can just be installed. That does not mean, we will never support it ;-)

I agree that an ECS or Filebeat package on it's own could make sense.

File: integration/README.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+{integration-name}/dataset/{dataset-name}/{package-structure}
+```
+
+On the top level of each integration, it contains the `manifest.yml`, `LICENSE.txt` file and the `changelog.yml`.
+
+## Assets
+
+Below are all the existing assets described. The assets which are already defined in [ASSET.md](https://github.com/elastic/integrations-registry/blob/master/ASSETS.md) from the package definition are not repeated here.
+
+### manifest.yml
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: I guess there are 2 manifests. The global ones and the ones per datasets. You think using the same name could become confusing?

File: integration/README.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+# The changelog.yml contains all the changes made to the integration and it's datasets.
+# If a dataset is adjusted, it should also be added to this changelog.
+# The changelog is in a structure format so the order does not matter and it can be used
+# for visualisation in the UI.
+
+  # Description of the change
+- description: Added dataset foo
+
+  # The versions here should follow semver
+  version: 1.2.3
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: SGTM, will probably also make it a bit more human readable.

File: integration/README.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+### docs/README.md
+
+README document which contains all the documentation about the integration. It is possible that each dataset has its own additional documentation. It is expected that this will be just appended to the main README on packaging.
+
+### img/
+
+Directory for all the icons, screenshots and potentially videos. 
+
+Question: Should we name this media?
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: long term, yes. Will rename.

File: integration/README.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+
+## Versioning
+
+The version of a package is taken from the manifest file. If the CoreDNS package contains `version: 1.2.3` it will build the package `coredns-1.2.3`. For now, no exact version of a dataset is specified. If a dataset is updated, next time packaging is called for an integration, it will pull in the newest assets. So if there is a breaking change in a dataset, it's up the integration package dev to decide if this is needed. To reduce errors we could introduce exact specification of a dataset version. This would mean in case a dataset version is updated, all datasets which reference it must be updated too. As everything is in one repository, this shouldn't be too much hassle but would make it more explicit.
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: I think the tricky part if something is in an other repository will be the `includes`. But I'm sure we could figure out some automation around copying these over.

File: integration/README.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+I expect most integrations to only be moving forward and have rarely breaking changes. Because of this no backports to branches etc. are needed. In case this is needed, there are several options:
+
+* Have a branch for this integration. Packaging will just work as is.
+* Have this integration in a separate repo.
+
+
+## Conversion of modules to integrations
+
+As the data of a module and an integration stay mostly the same, transformation from a module to an integration can mostly be automated. I started to play around with some tooling to convert existing modules to integrations but I would prefer to delay the discussion around this until we agreed on the format for building integrations.
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: I think the main thing this could change is that potentially there are more config files? But in the first phase I expect these are still shipped as part of MB. Only when we get to inputs, the configs the user has to past in become much bigger?

File: integration/apache/dataset/apache.access/dataset.yml [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        os.darwin:
+          - /usr/local/var/log/apache2/access_log*
+        os.windows:
+          - "C:/tools/Apache/httpd-2.*/Apache24/logs/access.log*"
+          - "C:/Program Files/Apache Software Foundation/Apache2.*/logs/access.log*"
+    
+        # TODO: Configure input for the module. This can be overwritten by central management.
+        input:
+
+ingest_pipeline: ingest/default.json
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/apache/dataset/apache.access/dataset.yml]: This will actually become obsolete. I'm planning to use handlebar or similar inside the templates which must reference the exact name of the file in the package. This will then replaced with the name and a prefix to match the template.

File: integration/apache/dataset/apache.access/dataset.yml [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+      default: ["%base_path/access.log"]
+      validations:
+        presence: true
+  transport.tcp:
+    port:
+      type: range # We can have Port type which has the default validations.
+      default: 8000
+      validations:
+        min: 0
+        max: 65568
```
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/apache/dataset/apache.access/dataset.yml]: We will actually not need this for our first version. This is a copy / paste from an initial proposal from @ph .

File: integration/apache/dataset/apache.access/testing.yml (hunk 3 above)
Comment (ruflin) [reply integration/apache/dataset/apache.access/testing.yml]: Yes, I think the testing.yml is the one the least fleshed out yet. I hope we can take over large parts from what we have today but also get new goodies like what the testing team is working on.

File: integration/apache/dataset/apache.access/testing.yml (hunk 2 above)
Comment (ruflin) [reply integration/apache/dataset/apache.access/testing.yml]: +1

File: integration/apache/dataset/apache.access/kibana/index-pattern/metricbeat.json (hunk 1 above)
Comment (ruflin) [reply integration/apache/dataset/apache.access/kibana/index-pattern/metricbeat.json]: Looks like a copy / paste mistake. I think in general we should still not ship any index patterns but they are generated on the fly based on the fields.yml like today.

File: integration/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: Yeah, ok, removing dependencies support from the installation can make sense, we may not need it for most of cases :+1:

File: integration/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: We may end up needing some other repository for build-time dependencies :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: Or maybe the external source can be specified in the include itself, e.g. for an integration including ECS:
```
includes:
- name: "ecs:ecs"
  source: https://github.com/elastic/integrations
```

To allow something like this in the future without needing to support transitive dependencies, we may add the restriction that datasets with `package: false` cannot have `includes`, and only datasets with `package: false` can be included.

And this makes me wonder if datasets with `package: false` should just be another thing like "libraries"(?) instead of datasets, as they probably will have more restrictions.

File: integration/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: Yes, current light modules can be shipped as part of MB, but as they are just plain text files we could have integrations with them.

File: integration/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: I am ok with this being named `manifest.yml` too, but the datasets manifest is now `dataset.yml`, I think that the confusion can come from this change, though this is probably not so important.

File: integration/README.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+The format of a package for the registry can be found here: https://github.com/elastic/integrations-registry#package-structure But for the development of a package, this is not enough as we also need testing of datasets which needs additional meta data. The proposed structure allows to tests metricsets / filsets in a similar way as we do today. With `mage package` all assets are packaged together to conform to the package structure.
+
+# Definitions
+
+**Integration package**: An integration package is a packaged version of an integration. This is what is served by the integrations registry. An example on what such a package looks like can be found here. It’s important to state that the shipped package does not look identical to the format here which is optimised for development and testing.
+
+**Integration**: Integration definition with manifest and several datasets with the assets for the Elastic Stack.
+
+**Dataset**: Group of assets which are grouped together for testing and development purposes.
```
Comment (sayden) [inline integration/README.md]: If this is a _group of assets_, why don't we call them "Assets"? In my head, when I read _datasets_ I'm thinking in some sort of structured information about some topic that have been already extracted from "somewhere" and is ready to analyze. The wikipedia definition also states something similar: https://en.wikipedia.org/wiki/Data_set

Just doing some "duck testing" here :smile:

File: integration/README.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+
+# Integration files structure
+
+As today with modules, the base structure for the implementation is split up into integrations which contains multiple datasets. All the assets are inside each dataset and the dataset file structure is very similar to the final structure of a package. The structure looks as following:
+
+```
+{integration-name}/dataset/{dataset-name}/{package-structure}
```
Comment (sayden) [inline integration/README.md]: I think it may help comprehension a lot to add a example file tree structure after this line.

File: integration/apache/dataset/apache.access/dataset.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/apache/dataset/apache.access/dataset.yml]: This may be interesting for that https://github.com/elastic/beats/pull/13357

File: integration/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: I just found the initial reason I went for a pure array. I initially thought that each dataset would contain its own changelog. Then combining the changelogs would have been just concatenating the files. and order would not have mattered. But as we have just 1 changelog per integration now and I don't expect us to import changelog entries from a dependency, this is not necessary anymore.

File: integration/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: All good questions :-)

I initially had it in its separate library directory. This had two disadvantages: 

* My collection script became more more complex
* It's not possible to have a dependency "directly" on an existing package. Let me give an example: I'm assuming ECS will become it's own package that we should also ship. But almost all integrations will depend on it. We could now move 99% of the package into library or have it as a package which others depend ont. Having it's own package solves also the changelog question for example and I would guess it's easier to understand / find. But technically I guess both options work.

So in summary, I think `package: false` cannot include others, but I don't think that `package: true` means it cannot be included. But I see how this could bring us in building include-trees. Perhaps we could go with this limitation to keep things simple to get started.

File: integration/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: I guess for now, we need to keep them shipped with MB. With Fleet, it will become easy to just ship them to the agent. I assume lightweight modules do not necessarly require a restart of MB? The other option is, that we let user copy / paste it as a backup.

File: integration/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: I'm on board with renaming `dataset.yml` to also `manifest.yml` if that is what you mean?

File: magefile.go (hunk 4 above)
Comment (ruflin) [reply magefile.go]: Agree, should be an external tool in the end that can also be used in any other repo. I don't think it should become part of the Beat because that is "our way" of developing integrations. If someone develops just 1 integration in a repository without dependencies, testing etc. he could just go with the package structure and version it. Building the package is creating a zip file.

File: integration/README.md (hunk 14 above)
Comment (ruflin) [reply integration/README.md]: The description here is unfortunate. These assets together are needed to create a dataset. But I think you have a point here and it needs either some revising of the description or potential renaming. More "duck testing" needed :-)

File: integration/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: @jsoriano Changing the code, I just had an idea for an other name: `imports` ?

File: integration/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: Add the definition of the above as the "standard" for our packages: https://github.com/elastic/integrations-registry/blob/master/ASSETS.md#changelogyml And pushed a commit to change it in the docs.

File: integration/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: Thinking about this again: The good news with the syntax we have (independent of the name), we can always extend it with more fields as it's an array. not having a source, just assumes it's local.

File: integration/apache/dataset/apache.access/dataset.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/apache/dataset/apache.access/dataset.yml]: I removed this line to prevent confusion.

File: integration/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: Looking at this again, renaming it to manifest.yml makes the most sense, as all the "sub" manifests contribute to the overall manifest in the end.

File: integration/apache/dataset/apache.access/kibana/index-pattern/metricbeat.json (hunk 1 above)
Comment (ruflin) [reply integration/apache/dataset/apache.access/kibana/index-pattern/metricbeat.json]: Move it to apache.stats for now, but we probably will not need it in the end.

File: integration/apache/dataset/apache.access/testing.yml (hunk 2 above)
Comment (ruflin) [reply integration/apache/dataset/apache.access/testing.yml]: Added a note to the file directly about this.

File: integration/README.md (hunk 15 above)
Comment (ruflin) [reply integration/README.md]: Updated the README with an example path.

File: integration/README.md (hunk 14 above)
Comment (ruflin) [reply integration/README.md]: I'm tempted to agree here that we should call the directory inside each integration `assets` instead of `dataset`. Because in some cases, it is literally only assets and nothing in it generates the dataset. One of the reasons of this grouping is also the "things" we test together. All these assets are in one place. It seems to be straight forward to explain that everything we need for testing together goes into one asset directory with a name. @sayden Would that help?

@jsoriano WDYT?

Comment (ruflin) [reply integration/README.md]: Trying to also add an an example here:

The coredns integration has 3 asset groups:

* coredns (all the generic assets)
* log (assets for the log dataset)
* stats (assets for the stats dataset)

File: integration/README.md [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+```
+
+An example of the above for an apache.access log ingest pipeline is:
+
+```
+apache/dataset/access/elasticsearch/ingest-pipeline/default.json
+```
+
+On the top level of each integration, it contains the `manifest.yml`, `LICENSE.txt` file and the `changelog.yml`.
```
Comment (ruflin) [inline integration/README.md]: @jsoriano I guess we don't have to add a `LICENSE.txt` file into each integration. We can add it during packaging time based on the entry in the `manifest.yml`.

File: integration/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: SGTM, I would only avoid the generic `datasets` :slightly_smiling_face:

File: integration/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: > I'm assuming ECS will become it's own package that we should also ship. But almost all integrations will depend on it.

But then if it is packaged it won't have `package: false`, right? Or I am misunderstanding something? :slightly_smiling_face:

File: integration/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: Yeah, light modules don't require a restart of MB, they are loaded "lazily" when instantiated, so new light modules can already be installed in run time.

File: integration/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply integration/README.md]: Yes, I meant that we keep the `manifest.yml` name for all datasets, or at least if we rename things we don't use this name for other things and not for its current usage :slightly_smiling_face: what could be confusing.

File: magefile.go (hunk 4 above)
Comment (jsoriano) [reply magefile.go]: Yeah, agree with rather doing it as an external tool.
With current differences between filesets for filebeat, metricsets for metricbeat and so on it could make sense to have the tooling on each beat, but not with integrations, that won't need to be coupled to any beat.

File: integration/README.md (hunk 14 above)
Comment (jsoriano) [reply integration/README.md]: Yes, on a first read `dataset` also sounded a bit funny to me for that. +1 to consider another naming like `assets`.
For the directory we could also think in something like `src` or `source` as they contain the "source code" of the package.

File: integration/README.md (hunk 16 above)
Comment (jsoriano) [reply integration/README.md]: I guess that we'd need a license per repository, and it should be added to the generate packages, yes.

File: magefile.go (hunk 4 above)
Comment (ruflin) [reply magefile.go]: I suggest we define it as a goal to have this binary separated but first build it as part of this repo.

File: integration/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply integration/README.md]: yes, it will not have `package: false` but others will still depend on it.

File: integration/README.md (hunk 14 above)
Comment (ruflin) [reply integration/README.md]: src reminds me too much of source code but it's actually mostly "config" files. My favorite at the moment: assets

File: integration/README.md (hunk 16 above)
Comment (ruflin) [reply integration/README.md]: We probably need more then 1 ;-)

File: magefile.go (hunk 4 above)
Comment (jsoriano) [reply magefile.go]: Yes, agree with having the source in this repo, but I'd still consider doing it as a standalone tool, that can be also used with integrations in other repos without copying magefiles, just running the resulting binary.

Comment (ruflin) [reply magefile.go]: Yes, agree. I would just not make the conversion of mage to a binary a "blocker" of this PR.

Comment (jsoriano) [reply magefile.go]: Ok

File: integration/README.md [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+# Overview
+
+This describes how integrations are built and then packaged for the [integrations registry](https://github.com/elastic/integrations-registry). The format described in this document is focused allowing build and test integrations which then are packaged with tools. 
+
+The format of a package for the registry can be found here: https://github.com/elastic/integrations-registry#package-structure But for the development of a package, this is not enough as we also need testing of datasets which needs additional meta data. The proposed structure allows to tests metricsets / filsets in a similar way as we do today. With `mage package` all assets are packaged together to conform to the package structure.
+
+# Definitions
+
+**Integration package**: An integration package is a packaged version of an integration. This is what is served by the integrations registry. An example on what such a package looks like can be found here. It’s important to state that the shipped package does not look identical to the format here which is optimised for development and testing.
```
Comment (cachedout) [inline integration/README.md]: It looks like the link to what a package looked like was stripped out.

File: integration/README.md [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+
+
+## Versioning
+
+The version of a package is taken from the manifest file. If the CoreDNS package contains `version: 1.2.3` it will build the package `coredns-1.2.3`. For now, no exact version of a dataset is specified. If a dataset is updated, next time packaging is called for an integration, it will pull in the newest assets. So if there is a breaking change in a dataset, it's up the integration package dev to decide if this is needed. To reduce errors we could introduce exact specification of a dataset version. This would mean in case a dataset version is updated, all datasets which reference it must be updated too. As everything is in one repository, this shouldn't be too much hassle but would make it more explicit.
```
Comment (cachedout) [inline integration/README.md]: Is the intention for this to be [semantic versioning](https://semver.org/)? If so, we should explicitly declare that, I think.

**Comment by ruflin** [comment]
I'm currently in the process of extending the package structure with inputs: https://github.com/elastic/integrations-registry/pull/110 This should simplify some of the discussions we have here and brings up much closer to what we have today with datasets.

**Comment by ruflin** [comment]
I had a conversation with @jsoriano about potential changes in the package structur (https://github.com/elastic/integrations-registry/pull/110) and how it affects this building of package. Overall it should simplify the building of packages as what is built is already almost identical to the end package format. We basically replace what we have today as `dataset` by `inputs`. 

The fields which will be used for the package should be identical. Config options which are needed for building the packages will go into `build.yml` in each integration and `tests` directory for the testing part. This will make building the package simple as files can be copied and these assets be removed.

One of the open questions is on what happens with the fields from the processors and autodiscovery. It would be nice if these mappings would not have to be copied to each input and be part of the template. Instead they should be loaded globally by the package manager. More thoughts needed on how this will work.

I will update this PR to be adjusted to the new format.

**Comment by mtojek** [comment]
@ruflin Can we close this draft PR?

**Comment by ruflin** [comment]
Yes, so many things have changed since I started this.
